build/cache/cache.go

```diff
@@ -0,0 +1,142 @@
+	path, err := os.UserCacheDir()
+	if err == nil {
+		return filepath.Join(path, "gopherjs", "build_cache")
+	}
+
+	return filepath.Join(build.Default.GOPATH, "pkg", "gopherjs_build_cache")
+}()
+
+// cachedPath returns a location inside the build cache for a given set of key
+// strings. The set of keys must uniquely identify cacheable object. Prefer
+// using more specific functions to ensure key consistency.
+func cachedPath(keys ...string) string {
+	key := path.Join(keys...)
+	if key == "" {
+		panic("CachedPath() must not be used with an empty string")
+	}
+	sum := fmt.Sprintf("%x", sha256.Sum256([]byte(key)))
+	return filepath.Join(cacheRoot, sum[0:2], sum)
+}
+
+// Clear the cache. This will remove *all* cached artifacts from *all* build
+// configurations.
+func Clear() error {
+	return os.RemoveAll(cacheRoot)
+}
+
+// BuildCache manages build artifacts that are cached for incremental builds.
+//
+// Cache is designed to be non-durable: any store and load errors are swallowed
+// and simply lead to a cache miss. The caller must be able to handle cache
+// misses. Nil pointer to BuildCache is valid and simply disables caching.
+//
+// BuildCache struct fields represent build parameters which change invalidates
+// the cache. For example, any artifacts that were cached for a minified build
+// must not be reused for a non-minified build. GopherJS version change also
+// invalidates the cache. It is callers responsibility to ensure that artifacts
+// passed the the StoreArchive function were generated with the same build
+// parameters as the cache is configured.
+//
+// There is no upper limit for the total cache size. It can be cleared
+// programmatically via the Clear() function, or the user can just delete the
+// directory if it grows too big.
+//
+// TODO(nevkontakte): changes in the input sources or dependencies doesn't
+// currently invalidate the cache. This is handled at the higher level by
+// checking cached archive timestamp against loaded package modification time.
+//
+// TODO(nevkontakte): this cache could benefit from checksum integrity checks.
+type BuildCache struct {
+	GOOS      string
+	GOARCH    string
+	GOROOT    string
+	GOPATH    string
+	BuildTags []string
+	Minify    bool
+	// When building for tests, import path of the package being tested. The
+	// package under test is built with *_test.go sources included, and since it
+	// may be imported by other packages in the binary we can't reuse the "normal"
+	// cache.
+	TestedPackage string
+}
+
+func (bc BuildCache) String() string {
+	return fmt.Sprintf("%#v", bc)
+}
+
+// StoreArchive compiled archive in the cache. Any error inside this method
+// will cause the cache not to be persisted.
+func (bc *BuildCache) StoreArchive(a *compiler.Archive) {
+	if bc == nil {
+		return // Caching is disabled.
+	}
+	path := cachedPath(bc.archiveKey(a.ImportPath))
+	if err := os.MkdirAll(filepath.Dir(path), 0750); err != nil {
+		return
+	}
+	f, err := os.Create(path)
```

Comment:
nit:  Should we open this file under a different/temporary name, and do an atomic* rename at the end?

I'm concerned that otherwise concurrent compilations (which may be quite common in certain IDE environments with compile/test-on-save type features) could unintentionally read a half-written file.

*where supported by the filesystem, anyway.

Reply:
Done.